### PR TITLE
Allow Connect API Key

### LINKED ISSUE #135: Throw error if RSC url has `#`
See https://github.com/rstudio/shinyloadtest/pull/133#issuecomment-696792714

If the url contains a `#` and the instance is RSC, we should throw an error stating they should use the solo url location.

cc @akgold

### PR DESCRIPTION
**Pull request by akgold**
Related https://github.com/rstudio/shinycannon/pull/56

Fixes #135

### CODE REVIEW

File: R/shiny-recorder.R [hunk 1]
```diff
@@ -239,12 +244,12 @@ RecordingSession <- R6::R6Class("RecordingSession",
...
-        if (is.null(username)) stop("Login aborted (username not provided)")
-        password <- getPass::getPass("Enter your password: ")
-        if (is.null(password)) stop("Login aborted (password not provided)")
-        cookies <- postLogin(private$targetURL, private$targetType, username, password)
+      if (isProtected(private$targetURL) && is.null(private$connectApiKey)) {
+          username <- getPass::getPass("Enter your username: ")
+          if (is.null(username)) stop("Login aborted (username not provided)")
+          password <- getPass::getPass("Enter your password: ")
+          if (is.null(password)) stop("Login aborted (password not provided)")
+          cookies <- postLogin(private$targetURL, private$targetType, username, password)
```
Comment (schloerke) [inline R/shiny-recorder.R:252]:
```suggestion
        username <- getPass::getPass("Enter your username: ")
        if (is.null(username)) stop("Login aborted (username not provided)")
        password <- getPass::getPass("Enter your password: ")
        if (is.null(password)) stop("Login aborted (password not provided)")
        cookies <- postLogin(private$targetURL, private$targetType, username, password)
```

File: R/shiny-recorder.R [hunk 2]
```diff
@@ -239,12 +244,12 @@ RecordingSession <- R6::R6Class("RecordingSession",
     },
     initializeSessionCookies = function() {
       cookies <- data.frame()
-      if (isProtected(private$targetURL)) {
-        username <- getPass::getPass("Enter your username: ")
-        if (is.null(username)) stop("Login aborted (username not provided)")
-        password <- getPass::getPass("Enter your password: ")
-        if (is.null(password)) stop("Login aborted (password not provided)")
-        cookies <- postLogin(private$targetURL, private$targetType, username, password)
+      if (isProtected(private$targetURL) && is.null(private$connectApiKey)) {
```
Comment (schloerke) [inline R/shiny-recorder.R:247]:
```suggestion
      # if auth is required and are not using the connectApiKey...
      if (isProtected(private$targetURL) && is.null(private$connectApiKey)) {
```

**Review by schloerke — APPROVED** [review]

**Review by wch — APPROVED** [review]

**Comment by akgold** [comment]
Was just demoing this to the SE team -- one thing popped up -- RStudio Connect provides 2 URLs for each content. There's the one that comes up in your URL bar when you navigate to the content inside the Connect dashboard, and the "content URL" for solo mode. 

If you put the former into `shinyloadtest` with an API key, the system just hangs indefinitely with no error. `shinycannon` does throw an error 
```
CONNECT_API_KEY=T5opptrM4yXfOlFwDQwta4qjedebPsQB java -jar target/shinycannon-1.1.0-jar-with-dependencies.jar ../shinyloadtest/recording.log https://colorado.rstudio.com/rsc/connect/#/apps/7278
zsh: bad pattern: https://colorado.rstudio.com/rsc/connect/#/apps/7278
```
But this error probably could be more friendly/helpful.

I think we should check for the form of the URL (look for a `#`?) and throw an error.

@schloerke -- see e.g. https://colorado.rstudio.com/rsc/connect/#/apps/7278/access vs https://colorado.rstudio.com/rsc/content/7278 (I've given you access to the app on Colorado RSC if you want to take a look). Want me to make a PR here? I'm less confident of my ability to do so in `shinycannon`.

**Comment by schloerke** [comment]
@akgold Issue added: https://github.com/rstudio/shinyloadtest/issues/135

File: R/shiny-recorder.R [hunk 3]
```diff
@@ -183,14 +183,18 @@ CLIENT_WS_STATE <- enum(UNOPENED, OPEN, CLOSED)
...
       private$targetType <- servedBy(private$targetURL)
       if (private$targetType == SERVER_TYPE$SAI) {
         stop("Recording shinyapps.io apps is not supported")
       }
       private$localHost <- host
       private$localPort <- port
+      if (private$targetType == SERVER_TYPE$RSC) {
+        message("Authenticating using provided connect_api_key.")
+        private$connectApiKey <- connectApiKey
+      }
```
Comment (schloerke) [inline R/shiny-recorder.R:197]:
```suggestion
      if (private$targetType == SERVER_TYPE$RSC && !is.null(connectApiKey)) {
        message("Authenticating using provided connect_api_key.")
        private$connectApiKey <- connectApiKey
      }
```

Comment (schloerke) [reply R/shiny-recorder.R:197]: This prevents a constant message of `"Authenticating using provided connect_api_key"` whenever user tests agains a RSC instance regardless if there is a key.

Now a key is required to produced the statement (and store it).

File: R/shiny-recorder.R [hunk 4]
```diff
@@ -183,14 +183,18 @@ CLIENT_WS_STATE <- enum(UNOPENED, OPEN, CLOSED)
...
       private$targetType <- servedBy(private$targetURL)
       if (private$targetType == SERVER_TYPE$SAI) {
         stop("Recording shinyapps.io apps is not supported")
       }
       private$localHost <- host
       private$localPort <- port
+      if (private$targetType == SERVER_TYPE$RSC && !is.null(connectApiKey)) {
+        message("Authenticating using provided connect_api_key.")
+        private$connectApiKey <- connectApiKey
+      }
```
Comment (schloerke) [inline R/shiny-recorder.R:197]:
```suggestion
      if (private$targetType == SERVER_TYPE$RSC) {
        if (!is.null(connectApiKey)) {
          message("Authenticating using provided connect_api_key.")
          private$connectApiKey <- connectApiKey
        }
        if (grepl("#", targetAppUrl, fixed = TRUE)) {
          stop("Please provide the content URL (solo mode) of this RStudio Connect Shiny app")
        }
      }
```

**Review by jcheng5 — APPROVED** [review]

**Review by jcheng5 — APPROVED** [review]

**Pull request merged**